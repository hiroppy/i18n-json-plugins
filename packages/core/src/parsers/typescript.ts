import {
  CallExpression,
  createSourceFile,
  forEachChild,
  ScriptTarget,
  SyntaxKind,
} from "typescript";
import type { Node } from "typescript";
import { Res } from "./types";

export function parse(
  code: string,
  defaultI18nFilename: string = "common"
): Res {
  const src = createSourceFile("code.tsx", code, ScriptTarget.Latest, true);
  const res: Res = {};

  walk(src);

  function initI18nFileField(path: string) {
    if (res[path] === undefined) {
      res[path] = {
        keys: [],
      };
    }
  }

  function pushKeyToRes(path: string, key: string) {
    const target = res[path];

    if (!target) {
      throw new Error("not found the key");
    }

    if (target.keys.includes(key)) {
      return;
    }

    target.keys.push(key);
  }

  function walk(node: Node) {
    if (node.kind === SyntaxKind.CallExpression) {
      const methodName: string =
        // @ts-expect-error
        (node as CallExpression).expression.escapedText;

      if (methodName === "useTranslation") {
        // @ts-expect-error
        if (node.arguments.length === 0) {
          // common
          initI18nFileField(defaultI18nFilename);
        } else {
          // @ts-expect-error
          const i18nPaths = node.arguments[0];

          if (i18nPaths.elements === undefined) {
            // useTranslation('foo')
            initI18nFileField(i18nPaths.text);
          } else {
            // useTranslation(['foo', 'bar'])
            for (const path of i18nPaths.elements) {
              initI18nFileField(path.text);
            }
          }
        }
      }

      // TODO: support an aliased variable
      if (methodName === "t") {
        // @ts-expect-error
        const key: string | undefined = node.arguments?.[0].text;

        if (!key) {
          throw new Error("not found the key");
        }

        const hasNamespace = key.includes(":");

        if (hasNamespace) {
          const [k, v] = key.split(":");

          pushKeyToRes(k, v);
        } else {
          if (Object.keys(res).length !== 1) {
            throw new Error();
          }

          pushKeyToRes(Object.keys(res)[0], key);
        }
      }
    }

    forEachChild(node, walk);
  }

  return res;
}
