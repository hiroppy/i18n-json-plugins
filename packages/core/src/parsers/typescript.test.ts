import { parse } from "./typescript";

describe("parse", () => {
  test("should collect keys from a common file", () => {
    const res = parse(
      `
const { t } = useTranslation();

console.log(t('foo.bar'));
console.log(t('foo.bar'));
console.log(t('foo.bar.baz'));
    `,
      "common"
    );

    expect(res).toMatchSnapshot();
  });

  test("should collect keys from 1 file", () => {
    const res = parse(`
const { t } = useTranslation('test');

console.log(t('foo.bar'));
console.log(t('foo.bar'));
console.log(t('test:foo.bar.baz'));
console.log(t('test:foo.bar.baz'));
    `);

    expect(res).toMatchSnapshot();
  });

  test("should collect keys from multiple files", () => {
    const res = parse(`
  const { t } = useTranslation(['test', 'test1']);

  console.log(t('test:foo'));
  console.log(t('test:foo.bar'));
  console.log(t('test:foo.bar'));
  console.log(t('test1:foo.bar'));
  console.log(t('test1:foo.bar.baz'));
  console.log(t('test1:foo.bar.baz'));
      `);

    expect(res).toMatchSnapshot();
  });

  test("should collect keys from jsx", () => {
    const res = parse(`
const Component = () => {
    const { t } = useTranslation(['test', 'test1']);

  return (
    <div>
      <p>{t('test:foo')}</p>
      <p>{t('test1:bar')}</p>
    </div>
  );
}
      `);

    expect(res).toMatchSnapshot();
  });
});
