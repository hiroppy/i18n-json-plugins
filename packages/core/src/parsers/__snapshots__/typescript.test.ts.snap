// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`parse should collect keys from 1 file 1`] = `
Object {
  "test": Object {
    "keys": Array [
      "foo.bar",
      "foo.bar.baz",
    ],
  },
}
`;

exports[`parse should collect keys from a common file 1`] = `
Object {
  "common": Object {
    "keys": Array [
      "foo.bar",
      "foo.bar.baz",
    ],
  },
}
`;

exports[`parse should collect keys from jsx 1`] = `
Object {
  "test": Object {
    "keys": Array [
      "foo",
    ],
  },
  "test1": Object {
    "keys": Array [
      "bar",
    ],
  },
}
`;

exports[`parse should collect keys from multiple files 1`] = `
Object {
  "test": Object {
    "keys": Array [
      "foo",
      "foo.bar",
    ],
  },
  "test1": Object {
    "keys": Array [
      "foo.bar",
      "foo.bar.baz",
    ],
  },
}
`;
